text/0249-v2-experiments.md

```diff
@@ -373,6 +367,358 @@ Disadvantages:
+```
+
+**If there are no breaking changes, are the APIs really experimental?**
+
+Given that preview APIs are safe to use, and no breaking changes will be
+introduced to them, we might consider not referring to them in any special way,
+and if needed, simply add a version to the API name. The first version of
+`grantWrite` will be named `grantWrite`, the second version will be named
+`grantWriteV1` and so on. While a preview API will not break, we should still
+use a naming scheme that convey its non-final nature for the following reasons:
+
+1. **Real-estate**: `grantWrite` is a much better name than `grantWritePre3` for
+   the final API.
+2. **Encourage feedback**: Declaring an API as "in preview" encourage the
+   community to supply feedback.
+3. **Setting the right expectations**: When an API in preview users are aware
+   that this is not the final version of the API, and its deprecation is
+   expected.
+
+### Modules lifecycle
+
+This section discuss adding a new modules to `aws-cdk-lib`.
+
+#### Alpha (optional)
+
+This stage is intended for modules that are in early development stage, they
+change rapidly and may incurs many breaking changes. These modules will be
+released separately from `aws-cdk-lib` and will have their own version, which
+allows for breaking changes in accordance to semver, e.g `0.x` or prerelease
+qualifiers.
+
+In order to release a module as alpha, the following conditions must be met:
+
+1. Module from `aws-cdk-lib` **can not** depend on it.
+2. Can not depend on any other alpha module\*\*.
+
+The purpose of the first condition is to prevent cyclic dependencies. To
+illustrate the purpose of the second condition lets look at an example. Assume
+we have an alpha module A, which depends on an alpha module B, graduating module
+A means it will be added to `aws-cdk-lib`, since module A depends on module B,
+it means that adding module A to `aws-cdk-lib` breaks condition 1 and creates a
+cyclic dependency. To prevent the cycle we will have to graduate B as well,
+which is not necessarily what we want. Additionally, if module B is an alpha
+module, module A will have to declare a fixed dependency on it, locking its
+consumers to the same version of module B - defeating the purpose of
+peerDependencies.
+
+> \*\* _There might be exceptions that will be considered on a per case basis_
+
+#### Beta (optional)
+
+In this stage modules are added to `aws-cdk-lib` with a preview suffix in the
+name of the module, e.g `aws-batch-beta-x`. Modules in this stage are not
+allowed to introduce any breaking changes to their API. Any non backward
+compatible change will be introduced via deprecation\*\*.
+
+> \*\* _The deprecation process will be discussed in the API previews
+> specification_
+
+#### Stable
+
+In this stage modules are added to `aws-cdk-lib` under their final name, e.g
+`aws-batch`. Addition of new APIs should follow the API previews specification.
+
+### Multiple alpha modules VS. single module
```

Comment:
Here is my on position in the "one unstable module vs many unstable modules" debate.

I am in favor of having a single module. I have just one reason for that position: it's simply less risky. We are taking on a huge amount of risk already by "betting the farm" on Option 6 that does something we've never done in the CDK before. Therefore, whichever option we can choose here that is less risky, I'm in favor of.

Why am I saying that Option 2 is less risky? Because it follows the same scheme as MonoCDK. With MonoCDK, we're basically saying that "L2s are special", and we need to release them together. If that is the case, I don't see why unstable L2s should be treated any differently than stable L2s.

You say that Option 3 has no downsides compared to Option 2, because unstable packages can easily depend on each other using `peerDependencies`. Well, not everyone uses `npm 7`. In fact, `npm 7` today ships only with unstable versions of Node (15). The vast, vast majority of our users are on `npm 6`. Which means they will have to manually fulfill the peer dependencies of unstable modules - a terrible experience.

But let's assume everyone uses `npm 7`. In my opinion, there are still potential pitfalls in this solution. Let's consider the following scenario. We have 2 unstable modules, `@aws-cdk-lib-experiments/aws-A` and `@aws-cdk-lib-experiments/aws-B`, and they peer depend on `0.10.0` of another unstable package, `@aws-cdk-lib-experiments/aws-X`. Our customer (who writes a CDK application, not a library) uses `A` and `B` in their `package.json`, both in version `0.10.0`.

Some time passes, and more versions are released. Our customer upgrades `A` to version `0.20.0`, because they need some new features of it. That newer version of `A` depends on version `0.20.0` of `X`, which had some breaking changes made to it compared to version `0.10.0` of `X`.

What happens when a customer upgrades `A` in their `package.json` to `0.20.0`, while leaving `B` unchanged, on `0.10.0`? Will they get an error during `npm install`? Will they have two versions of `X` in their dependency closure now, risking getting the dreaded "Construct is not an instance of Construct" error? Will they get a runtime error from `B`, because version `0.10.0` of `B` does not work with version `0.20.0` of `X` (because of the breaking changes in `X`)? Will they be forced to upgrade `B` to `0.20.0` too, negating your stated benefit of this approach? What if there's no `0.20.0` of `B`, because it's released from a completely different repository, on a different schedule, than the other modules?

This is just one potential scenario. I'm sure there are many other ones that we haven't even thought of. And that to me is the definition of risk: the fact that we have these blind spots that may surprise us after we make this decision.

We already have experience that many single modules cause dependency problems in the NPM ecosystem - even for stable modules, because of the "Construct is not an instance of Construct" TypeScript error. Why would we go back on all that experience from `v1`, and decide to keep releasing multiple modules in `v2`? It seems incongruent to me to insist on MonoCDK for stable modules, but separate packages for unstable ones. Either we want MonoCDK (and if  so, Option 2 is less risky), or we want Option 3, and then we don't need MonoCDK (because of `npm 7` and peer depedencies).

We are making a huge bet with v2 being shipped as MonoCDK. We are saying that we fixed the dependency problem for the CDK ecosystem with this change. It has some downsides, like the huge package size, but we're claiming that the downsides are worth it, because the dependency issues will be addressed once and for all. If we ship MonoCDK v2, but still have the dependency hell problem, just this time restricted to unstable modules, I think it will be a disastrous outcome.

> If we release all modules under a single package, in which every release may include breaking changes to any numbers of its modules, users are force to accept the breaking changes in all modules, even if they only want to upgrade a single module.

I agree that is a downside to Option 2. All of the options have downsides, none of them are perfect, including Option 3. _However_, I'm willing to swallow this known inconvenience of Option 2, if the alternative is to risk all of the unknown consequences of Option 3.

Reply:
Version mismatch when depending on unstable modules will happen with a single or multiple packages, its just a matter of "where". If a use relies on a third party package that relies on aws-cdk-experimental v0.1 and additionally needs aws-cdk-experimental v0.2 themselves for some newer feature, this presents a mismatch. I think part of my inclination for leaning towards separate packages is that the process of resolving these is the same in both. Any packages that rely on unstable construct libraries (third party or otherwise) will need to consistently update the versions of these dependencies in order for usage of the "experimental" ecosystem to be viable. Especially across third party construct libraries. Making aws-cdk-lib never break is what allows monocdk to circumvent this not the bundling itself.

Making everything peer dependencies all the way down also affects this because there only ever should be a single version in the resolved node_modules of any app consuming this chain of packages. Conflicts ultimately have to be resolved by the user if they exist. NPM 7 makes this easier by default but the end product is essentially the same, the user can declare the ultimate version of any peer dependencies.

Reply:
Libraries cannot depend on unstable modules - that is a problem that is simply impossible to solve. We have learned this from `v1`.

Note that my scenario happens for an **application** customer, not a **library** customer.

Reply:
You are ignoring a few major differences:

1. **In v2 we will only be using peerDependencies:** The majority of problems in v1 were rooted in the fact that we declared peerDependencies **and** direct dependencies between modules. We did it because we had hundreds of modules, and adding them as peerDependencies would have forced consumers to install all of these modules themselves. **In v2 we will only use peerDependencies,** so even if a consumer will not use npm7, they will control the installed version. 
2. Due to the fact that experimental modules can introduce a breaking change in every release, libraries, including other experimental modules shouldn't depend on them.
3. From the 35 experimental modules we currently have - 24 (!) have no dependents and only stable dependencies. This means that there is no reason to treat them the same as third party libraries. If we can't make it work for us, we will never be able to grow the ecosystem.
 

Having a single package that includes all modules will also hurt the adoption of new experimental modules. If a new module is added in a specific version, which also includes breaking changes to other modules, customers might not being willing to upgrade if it means it will impact already deployed infrastructure. TThe fact that a breaking change in a cdk module may be a change to infrastructure that has already been deployed (or even downtime), means that customers are more hesitant to update their version. Having separate module mean we will reduce the blast radius of every update.

Reply:
>Note that my scenario happens for an application customer, not a library customer.

Totally. In your scenario with npm <7, the user will be prompted with an error saying that one or the other of their dependencies is missing a peer dependency. If both A and B libraries happen to be compatible with the version of X that is installed, this could be ignored and the app will work. If not, they will either need to explicitly install a version of X that they know to be compatible or also update the version of B as you suggested. This, as you mentioned, negates the positive effect of only needing to upgrade specific experimental packages when needed. Maybe we can do some analysis to get an idea of how often we could expect this to occur

I guess I'm unsatisfied with "Libraries cannot depend on unstable modules" because I believe they undoubtedly will and should be able to provided they are willing to take on that maintenance cost of keeping dependencies up to date. With that in mind separate experimental modules feel a lot like any other library that will rely on aws-cdk-lib during the 0.x phase of development. Some things you depend on may be stable, others may have unstable features that are of high value that you decide to use but with the known cost that 0.x comes with.

Reply:
Some of these points have already been addressed in comments by Mitch and Neta, but I'm going to add my own view on them.

## Dependencies between experimental packages, what if a user upgrades one but forgets the other?

First of all, I have a suspicion this will be **rare**¹, so I'm sure it's worth optimizing for. But even then:

> Will they have two versions of X in their dependency closure now,

No! I get you've been traumatized by this, but again:

**The problem of v1 was not modularity, the problem of v2 was using `dependencies` where we should have been using `peerDependencies`**

`dependencies` will lead to multiple copies of a library, `peerDependencies` will not. You will get an installation error if you match incorrectly.

If we had used `peerDependencies` in v1, we would NOT have been in **technical** problems. The reason we're choosing `monocdk` over modular packaging is a **usability** issue (user needing to specify all transitive deps), which has been solved in NPM7. Had NPM7 come out earlier, we probably wouldn't have done this at all (or maybe we would have, but we definitely would have had fewer complaints).

But NPM moves at the pace it does, and c'est la vie.

But conclusion:

**Modular packaging is not a technical risk, it's only a usability issue in case of [many] transitive dependencies for NPM <7.**

> *¹) I'm not saying they won't happen, I'm saying they will be rare. A single counterexample does not invalidate this argument.*

## Backwards compatibility and ecosystem

> We also know from v1 experience that you cannot build an ecosystem of packages if those packages do not maintain backwards compatibility.

It is true, but this is completely independent from packaging. In fact, I would put it like this:

**If we decide to break APIs between releases, then monopackaging is worse than granular packaging**

That's because you don't get to pick and choose versions anymore.

Let's introduce some notation so that we can be precise about this.We have APIs `A` and `B` and they have backwards incompatible revisions (`A1`, `A2`, `A3`, ...). Furthermore, let's say that if we leave old copies of the API in place in new releases of a library (but `@deprecated`), we're going to write that as `(A1) + A2` (in this case, `A1` is deprecated but still in the library, and `A2` is also in there²). 

Let's say these are the releases of `cdk-experiments-change`:

| `cdk-experiments-change@0.1.0` | `cdk-experiments-change@0.2.0` | `cdk-experiments-change@0.3.0` |
|--------|-----------|-----------|
| `A1 + B1` | `A2 + B1` | `A2 + B2` |

First we change `A`, then we change `B`.

Now, assume that we have an application or library, doesn't really matter which, that's happily consuming `A1`. They are on library version `0.1.0`, and all is swell.

Now `B2` just got a feature that they want to use while staying on `A1` for the `A` API but... that's impossible! There's no package version that supplies `A1 + B2`.

And this is *potentially* fine, you can just update your code to migrate from `A1` -> `A2` instead as well, right? No big deal.

It becomes more hairy once we start mixing in libraries that you may not have control over. Different scenario:

Maybe you are an application that uses a library that hard-locked you to `0.1.0` because it uses `B1` (and you can't know whether `0.2.0` will still have `B1` or not, so you have to minor-version-lock). But now the app wants to use `A2`. We're in the scenario that there's there *is* a version that supplies `A2 + B1`, and yet you're **still not able to take it** because your library dependency locked you into a version of the CDK lib.

```
app —[ uses L1 ]⟶ lib@1.0.0 —[ uses B1 ]⟶ cdk-experiments-change@0.1.0   
app —[ uses B2 ]⟶ cdk-experiments-change@0.2.0                    **** CONFLICT ****
```

Bottom line:

**Building ecosystems around in-place broken APIs get *worse* if we tie those APIs together in releases.**

The only way around this is to do one of two things:

### Don't in-place break APIs

| `cdk-experiments-add@0.1.0` | `cdk-experiments-add@0.2.0` | `cdk-experiments-add@0.3.0` |
|--------|-----------|-----------|
| `A1 + B1` | `(A1) + A2 + B1` | `(A1) + A2 + (B1) + B2` |

So now everyone can take upwards unbounded (caret) dependency ranges and everyone can move forward. This is option 6 and what we're doing for the **stable** core lib.

### Don't tie breaking changes together together

| `a-experiment-change@0.1.0` | `a-experiment-change@0.2.0` |
|--------|-----------|-----------|
| `A1` | `A2` |

| `b-experiment-change@0.1.0` | `b-experiment-change@0.2.0` |
|--------|-----------|-----------|
| `B1` | `B2` |

You still cannot  take range dependencies, but at least now you can mix and match dependencies on `a-experiment-change` and `b-experiment-change` at will without them stepping on each other.

> *² Let's disregard naming for a moment, but writing `(A1) + A2` will obviously have to mean that we'll have to have renamed some members in order to let those incompatible APIs exist in the same library side-by-side. The methods/classes of `A2` will have to have new names.*

## Unknown risk

> the risk all of the unknown consequences of Option 3.

See... this might be the crux of the issue. I don't see any unknown consequences that we haven't already encountered here, and I also don't see any avenue for "unknown unknowns" to crop up.

We've been vending granular libraries for ages now, and we know exactly what's up. There's three big issues with them:

* Multiple copies of a library ⟶ fixed by using `peerDependencies`
* Many dependencies is a chore for users ⟶ fixed by using NPM7 or not having many dependencies (likely!)
* Adding peerDependencies is a breaking change ⟶ fixed by using NPM7 or bumping your own MV (we do that anyway!)

They all seem perfectly well addressed, or not a big deal, and definitely only impacting a minority of users (so honestly I feel these are issue are not really worth all the air time we're giving them).

Between the two solutions: monoexperiments and granular experiments, the granular experiments option seems the "technically correct" alternative of the two. It also has the advantage that it will be easier for 3rd parties to contribute modules, which we can first incubate outside our repo, then inside our repo, then graduate to beta or higher.

Reply:
> The reason we're choosing monocdk over modular packaging is a usability issue

A logical follow-up question at this point would be: if mono-packaging has better usability, why don't we do the same for mono-experiments then?

We could. But in order to get the touted benefits of mono-packaging, we must have **additive** (non-breaking) APIs.

The whole point of having our experiments separately is so that we wouldn't have to do that and we could freely break APIs. If we want to break APIs, then mono-packaging actually produces (technical) downsides.

Reply:
So, to sum up the very long discussion (I'm guilty if this too):

Options 2 and 3 are very similar. All of the options involve some tradeoff, none of them are perfect. The tradeoff between 2 and 3 is the following:
* Option 2 allows the unstable modules to depend on each other (because it bundles them together), which option 3 does not allow.
* _However_, that bundling has the side-effect that all unstable modules are versioned together. So, if a customer wants to use 2 unstable modules, `A` and `B`, they always have to share a version, they can't use `A` in `0.10`, and `B` in `0.20`. This is the advantage that Option 3 has.

Now, it's a question of "which of the 2: unstable-to-unstable dependencies vs. separately-versioned unstables is more important?". I think reasonable people can disagree here. Neta and Rico (and probably Mitch) consider that "separate unstable versions" are more important. I consider "unstable-to-unstable" dependencies more important.

My particular reason (I don't want to speak for anyone else) for preferring "unstable-to-unstable" dependencies is that we know we need them with the current modules, because we have "unstable-to-unstable" dependencies _right now_. Which also means we will need this in the future, as AWS releases more services and more integrations between them.

As for the "separate versions of unstable", this is the situation today with `v1` and MonoCDK. And I've never heard customers saying that this particular limitation is causing them problems. So for me, it's "actual benefit we know we need right now" vs "possible downside that has not been reportd". Hence why I prefer Option 2.

But again, I think reasonable and smart people can disagree on this 🙂.

Here's the current snapshot of "unstable-to-unstable" dependencies (note that I removed a bunch of them relating to the fact that `cxapi`, `region-info`, `s3-assets`, etc. are still unstable - I assume those will become stable with V2). I'll let everyone be the judge if this is rare enough that it negates the downside of Option 3.

```
ℹ️️  Unstable package '@aws-cdk/aws-apigatewayv2-integrations' depends on unstable package '@aws-cdk/aws-apigatewayv2'
ℹ️️  Unstable package '@aws-cdk/aws-appmesh' depends on unstable package '@aws-cdk/aws-acmpca'
ℹ️️  Unstable package '@aws-cdk/aws-backup' depends on unstable package '@aws-cdk/aws-efs'
ℹ️️  Unstable package '@aws-cdk/aws-docdb' depends on unstable package '@aws-cdk/aws-efs'
ℹ️️  Unstable package '@aws-cdk/aws-eks' depends on unstable package '@aws-cdk/lambda-layer-awscli'
ℹ️️  Unstable package '@aws-cdk/aws-eks' depends on unstable package '@aws-cdk/lambda-layer-kubectl'
ℹ️️  Unstable package '@aws-cdk/aws-s3-deployment' depends on unstable package '@aws-cdk/lambda-layer-awscli'
ℹ️️  Unstable package '@aws-cdk/aws-ses-actions' depends on unstable package '@aws-cdk/aws-ses'
```

Reply:
We have solid arguments arguments for each option, but its getting really hard to follow the conversation in this PR.

I'm going to summarize the major pro and cons of every approach **when compared to each other** in a short doc, we will review it on Monday and make a decision.

Reply:
@skinny85 option 3 does allow unstable module to depend on other unstable modules (with the same graduation limitation as option 2), see Rico's response to your  "Will they have two versions of X in their dependency closure now" comment.